GVFS/GVFS/CommandLine/HealthVerb.cs

```diff
@@ -0,0 +1,226 @@
+                    {
+                        this.Output.WriteLine("Bad response from modified path pipe: " + modifiedPathsResponse.Header);
+                        return;
+                    }
+
+                    modifiedPathsList = modifiedPathsResponse.Body.Split(new char[] { '\0' }, StringSplitOptions.RemoveEmptyEntries);
+                }
+                catch (BrokenPipeException e)
+                {
+                    this.ReportErrorAndExit("Unable to communicate with GVFS: " + e.ToString());
+                }
+
+                foreach (string path in modifiedPathsList)
+                {
+                    if (path.Last() == GVFSConstants.GitPathSeparator)
+                    {
+                        path.TrimEnd(GVFSConstants.GitPathSeparator);
+                        pathData.ModifiedFolderPaths.Add(path);
+                    }
+                    else
+                    {
+                        pathData.ModifiedFilePaths.Add(path);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get two lists of placeholders, one containing the files and the other the directories
+        /// Goes to the SQLite database for the placeholder lists
+        /// </summary>
+        /// <param name="enlistment">The current GVFS enlistment being operated on</param>
+        /// <param name="filePlaceholders">Out parameter where the list of file placeholders will end up</param>
+        /// <param name="folderPlaceholders">Out parameter where the list of folder placeholders will end up</param>
+        private void GetPlaceholdersFromDatabase(GVFSEnlistment enlistment, GVFSEnlistmentHealthCalculator.GVFSEnlistmentPathData pathData)
+        {
+            List<IPlaceholderData> filePlaceholders = new List<IPlaceholderData>();
+            List<IPlaceholderData> folderPlaceholders = new List<IPlaceholderData>();
+
+            using (GVFSDatabase database = new GVFSDatabase(new PhysicalFileSystem(), enlistment.EnlistmentRoot, new SqliteDatabase()))
+            {
+                PlaceholderTable placeholderTable = new PlaceholderTable(database);
+                placeholderTable.GetAllEntries(out filePlaceholders, out folderPlaceholders);
+            }
+
+            pathData.PlaceholderFilePaths.AddRange(filePlaceholders.Select(placeholderData => placeholderData.Path));
+            pathData.PlaceholderFolderPaths.AddRange(folderPlaceholders.Select(placeholderData => placeholderData.Path));
+        }
+
+        private void GetPathsFromGitIndex(GVFSEnlistment enlistment, GVFSEnlistmentHealthCalculator.GVFSEnlistmentPathData pathData)
+        {
+            List<string> skipWorktreeFiles = new List<string>();
+            GitProcess gitProcess = new GitProcess(enlistment);
+
+            GitProcess.Result fileResult = gitProcess.LsFiles(
+                line =>
+                {
+                    if (line.First() == 'S')

```

Comment:
Skip worktree files are not the same as modified paths.  You are adding to these to the modified paths list.  It is the paths that do not have the skip worktree bit that would be the ones that are off because of the modified paths list.

Reply:
Oh duh, yes. I'll switch that to say `line.First() == 'H'`

Reply:
Is there a test we can add for this as well (maybe we should look at including a functional test)?

Reply:
I think it would need to be a functional test, since the verb is what takes care of this logic.

Reply:
I think we should definitely have some functional tests for this before it gets merged in.  Also I would rename these paths to something like `gitTrackingPaths` because they are not paths with the skipworktree bit set.

Reply:
Also have you ran this against the big repos to see what the performance and memory usage is like?

Reply:
That name makes sense. And yeah I've ran it against the OS repo. It takes a little over 18 seconds to run on average, I haven't looked at memory usage just yet. The next step I think is definitely functional tests

Reply:
>It takes a little over 18 seconds to run on average

With how many placeholders and paths affected by the modified paths?

Reply:
Not many (like 0.01%) but far and away the slowest operations appear to be those that consume the git index in some way, so the ls-tree call and the ls-files call. Given these running conditions, the process builds to take up 811Mb of memory

Reply:
After going from 198 placeholders to 12,000 the run time was in the same range (18.2 seconds to 18.8) and the memory usage only increased by 5MB. While these aren't great to begin with, I think they are enough for the initial release and are something that can be revisited after for improvement.